### PR TITLE
Consolidate exceptions thrown out by different clients

### PR DESCRIPTION
**Pull request by brycewang-microsoft**
Consolidate the following types of exception thrown out by different clients, and update doc comments for public APIs and relevant tests.
+ `IOException`
+ `SocketException`
+ `WebSocketException`
+ `UnauthorizedAccessException`
+ `SwaggerException`
+ `HttpRequestException`
+ `IndexOutOfRangeException`
+ `CryptographicException`

### CODE REVIEW

File: iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs [hunk 1]
```diff
@@ -32,20 +32,20 @@ public abstract class AuthenticationWithTokenRefresh : IAuthenticationMethod
         /// The time buffer before expiry when the token should be renewed, expressed as a percentage of the time to live.
         /// The default behavior is that the token will be renewed when it has <see cref="DefaultSasRenewalBufferPercentage"/> percent or less of its lifespan left.
         ///</param>
-        /// <exception cref="ArgumentOutOfRangeException">Thrown if <paramref name="suggestedTimeToLive"/> is a negative timespan, or if
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs:35]: I could see an argument for keeping this type of exception. It is more specific.

Comment (drwill-ms) [reply iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs:35]: It inherits from ArgumentException, btw.

File: iothub/service/src/Twin/Models/TwinCollection.cs [hunk 2]
```diff
@@ -160,7 +160,7 @@ public TwinMetadata GetMetadata()
         /// Gets the last updated time for this property.
         /// </summary>
         /// <returns>Date-time instance representing the last updated time for this property.</returns>
-        /// <exception cref="System.NullReferenceException">Thrown when the metadata object is null.
+        /// <exception cref="NullReferenceException">Thrown when the metadata object is null.
```
Comment (drwill-ms) [inline iothub/service/src/Twin/Models/TwinCollection.cs:163]: This was a bug we couldn't fix, but we should fix now. Not in this PR, but we should fix it so we don't return an NRE.

**Review by drwill-ms — APPROVED** [review]
I agree with all the changes except removing ArgumentOutOfRangeException, but I wouldn't die on that hill.

File: iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs (hunk 1 above)
Comment (brycewang-microsoft) [reply iothub/device/src/Authentication/AuthenticationWithTokenRefresh.cs:35]: Yeah, throwing more specifically is better than maintaining a shorter list of exceptions we have in such a case. Sure, I will change it back.

**Review by tmahmood-microsoft — APPROVED** [review]

File: iothub/service/src/Twin/Models/TwinCollection.cs (hunk 2 above)
Comment (drwill-ms) [reply iothub/service/src/Twin/Models/TwinCollection.cs:163]: An SDK should never throw a NullReferenceException. That is a bug.

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/src/Authentication/HsmAuthentication/Transport/HttpRequestResponseSerializer.cs [hunk 3]
```diff
@@ -120,9 +120,17 @@ internal static async Task<HttpResponseMessage> DeserializeResponseAsync(HttpBuf
...
+                            throw new IotHubClientException($"Header value is invalid for {headerName}.", IotHubClientErrorCode.NetworkErrors);
+                        }
+
+                        try
+                        {
+                            await httpResponse.Content.LoadIntoBufferAsync(contentLength).ConfigureAwait(false);
+                        }
+                        catch (HttpRequestException ex)
+                        {
+                            throw new IotHubClientException(ex.Message, IotHubClientErrorCode.NetworkErrors, ex);
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/HsmAuthentication/Transport/HttpRequestResponseSerializer.cs:132]: is transient?

Comment (drwill-ms) [reply iothub/device/src/Authentication/HsmAuthentication/Transport/HttpRequestResponseSerializer.cs:132]: Implied from network errors?

File: iothub/service/src/Utilities/Argument.cs [hunk 4]
```diff
@@ -120,13 +120,13 @@ internal static void ValidateBufferBounds(byte[] buffer, int offset, int size)
 
             if (offset < 0 || offset > buffer.Length || size <= 0)
             {
-                throw new ArgumentOutOfRangeException(nameof(offset));
+                throw new ArgumentOutOfRangeException(nameof(offset), "The buffer bound is invalid.");
```
Comment (drwill-ms) [inline iothub/service/src/Utilities/Argument.cs:123]: buffer bounds are invalid

File: iothub/device/src/Authentication/HsmAuthentication/Transport/HttpRequestResponseSerializer.cs (hunk 3 above)
Comment (brycewang-microsoft) [reply iothub/device/src/Authentication/HsmAuthentication/Transport/HttpRequestResponseSerializer.cs:132]: Yes, `IotHubClientErrorCode.NetworkErrors` will be taken as transient within ctor of `IotHubClientException`

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**